### PR TITLE
feature: add isFixedHeight parameter support for the JumpTo method of GridView

### PR DESCRIPTION
**Pull request by percival888**
add a feature support for GridView when it jump to specified index, now we can set isFixedHeight with true.

for GridView items of fixed height, there's no longer need to calculate with sliding, but calculates height and spacing, directly jumping to the specified position

### CODE REVIEW

File: lib/src/common/observer_controller.dart [hunk 1]
```diff
@@ -501,34 +501,125 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
-      final nextChildSize = isHorizontal
-          ? nextChildPaintBounds.width
-          : nextChildPaintBounds.height;
-      separatorTotalHeight = index * nextChildSize;
-    }
     final childPaintBounds = targetChild.paintBounds;
-    final childSize =
+    double itemSeparatorHeight = 0;
+    int indexLineCount = index;
+    if (obj is RenderSliverList || obj is RenderSliverFixedExtentList) {
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:507]: This judgment will cause problems with invalidate the jump of the third-party package's ListView.

File: lib/src/common/observer_controller.dart [hunk 2]
```diff
@@ -501,34 +501,125 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
+        if (isHorizontal) {
+          itemSeparatorHeight =
+              (nextChildOrigin.dx - targetChildOrigin.dx).abs() -
+                  childPaintBounds.width;
+        } else {
+          itemSeparatorHeight =
+              (nextChildOrigin.dy - targetChildOrigin.dy).abs() -
+                  childPaintBounds.height;
+        }
+      } else {
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:556]: Convenient for subsequent maintenance and reading. Can you add a commnet here?

File: lib/src/common/observer_controller.dart [hunk 3]
```diff
@@ -501,34 +501,125 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
+        var previousChild = obj.childBefore(targetChild);
+        var previousChildOrigin =
+            previousChild?.localToGlobal(Offset.zero) ?? Offset.zero;
+        while (previousChild != null &&
+            (isHorizontal
+                ? previousChildOrigin.dx == targetChildOrigin.dx
+                : previousChildOrigin.dy == targetChildOrigin.dy)) {
+          // calculate crossAxisSpacing
+          if (!isHaveSetCrossAxisSpacing) {
+            double crossAxisOriginX1 =
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:566]: I think the names of `crossAxisOriginX1` and `crossAxisOriginX2` should be changed. It does not only refer to the x-axis.

File: lib/src/common/observer_controller.dart [hunk 4]
```diff
@@ -501,34 +501,125 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
+                (targetChildOrigin.dy - previousChildOrigin.dy).abs() -
+                    childPaintBounds.height;
+          }
+        }
+      }
+      final childCrossAxisSize =
+          isHorizontal ? childPaintBounds.height : childPaintBounds.width;
+      int itemsPerLine = ((obj.constraints.crossAxisExtent + crossAxisSpacing) /
+              (childCrossAxisSize + crossAxisSpacing))
+          .round();
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:603]: Can you add a comment here explaining the purpose of `around` and `floor`?

File: lib/src/common/observer_controller.dart (hunk 1 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:507]: ok，I will improve this piece

File: lib/src/common/observer_controller.dart (hunk 2 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:556]: copy that

File: lib/src/common/observer_controller.dart (hunk 3 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:566]: you are right, i'll update the names

File: example/lib/features/gridview/gridview_fixed_height_demo/gridView_fixed_height_demo_page.dart [hunk 5]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (LinXunFeng) [inline example/lib/features/gridview/gridview_fixed_height_demo/gridView_fixed_height_demo_page.dart:1]: The file name 'gridView_fixed_height_demo_page.dart' isn't a snake_case identifier.
Try changing the name to follow the snake_case style.

File: example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                'item: $index',
+                style: const TextStyle(fontSize: 20, color: Colors.white),
+              ));
+        },
+        childCount: childCount,
+      ),
+    );
+  }
+
+  Widget _buildSeparator(double size) {
```
Comment (LinXunFeng) [inline example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart:121]: The declaration '_buildSeparator' isn't referenced.

File: example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        (BuildContext context, int index) {
+          if (grid1Context != context) grid1Context = context;
+          return Container(
+              height: 50,
+              color: Colors.primaries[index % Colors.primaries.length],
+              alignment: Alignment.topLeft,
+              child: Text(
+                'item: $index',
+                style: const TextStyle(fontSize: 20, color: Colors.white),
+              ));
```
Comment (LinXunFeng) [inline example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart:114]: It is recommended that you put a comma after the first parenthesis and format it.

File: lib/src/common/observer_controller.dart [hunk 8]
```diff
@@ -277,21 +277,23 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
   /// Jump to the specified index position without animation.
   ///
   /// If the height of the child widget and the height of the separator are
-  /// fixed, please pass the [isFixedHeight] parameter.
+  /// fixed, please pass the [isFixedHeight] parameter and the
+  /// [renderSliverType] parameter.
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:281]: It is recommended to explain the default handling when renderSliverType is not passed.

File: lib/src/common/observer_controller.dart [hunk 9]
```diff
@@ -277,21 +277,23 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
   ///
   /// If you do not pass the [isFixedHeight] parameter, the package will
   /// automatically gradually scroll around the target location before
   /// locating, which will produce an animation.
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:285]: Could you please add a description of the renderSliverType parameter?

File: lib/src/common/observer_controller.dart [hunk 10]
```diff
@@ -501,34 +510,140 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
-      final nextChildSize = isHorizontal
-          ? nextChildPaintBounds.width
-          : nextChildPaintBounds.height;
-      separatorTotalHeight = index * nextChildSize;
-    }
     final childPaintBounds = targetChild.paintBounds;
-    final childSize =
+    double itemSeparatorHeight = 0;
+    int indexOfLine = index;
+    if (obj is RenderSliverList ||
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:516]: The processing logic here is too long, it is recommended to split out the corresponding two methods and return the data in a model.

File: lib/src/common/observer_controller.dart [hunk 11]
```diff
@@ -501,34 +510,140 @@ mixin ObserverControllerForScroll on ObserverControllerForInfo {
...
+      nextChild ??= obj.childBefore(targetChild);
+      if (nextChild != null && nextChild is! RenderIndexedSemantics) {
+        // It is separator
+        final nextChildPaintBounds = nextChild.paintBounds;
+        itemSeparatorHeight = isHorizontal
+            ? nextChildPaintBounds.width
+            : nextChildPaintBounds.height;
+      }
+    } else if (obj is RenderSliverGrid ||
+        renderSliverType == ObserverRenderSliverType.grid) {
```
Comment (LinXunFeng) [inline lib/src/common/observer_controller.dart:530]: The processing logic of `_handleScrollToIndexForFixedHeight` method is too long, it is recommended to split out the corresponding two methods and return the data in a model.

File: example/lib/features/gridview/gridview_fixed_height_demo/gridView_fixed_height_demo_page.dart (hunk 5 above)
Comment (percival888) [reply example/lib/features/gridview/gridview_fixed_height_demo/gridView_fixed_height_demo_page.dart:1]: okay, i'll change the name to ‘grid_view_fixed_height_demo_page’

File: example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart (hunk 6 above)
Comment (percival888) [reply example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart:121]: Deleted

File: example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart (hunk 7 above)
Comment (percival888) [reply example/lib/features/scene/waterfall_flow_fixed_height_demo/waterfall_flow_fixed_height_page.dart:114]: okay

File: lib/src/common/observer_controller.dart (hunk 8 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:281]: copy that

File: lib/src/common/observer_controller.dart (hunk 9 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:285]: copy that

File: lib/src/common/observer_controller.dart (hunk 11 above)
Comment (percival888) [reply lib/src/common/observer_controller.dart:530]: yes, it would be better

**Review by LinXunFeng — APPROVED** [review]
LGTM, Thanks for your contribution.

**Pull request merged**